<?xml version="1.0"?>
<launch>
    <arg name="roomba"      default="roomba1" />
    <arg name="wait_time"   default="0.0" />
    <arg name="date"        default="20220224" />
    <arg name="config_path" default="$(find object_localizer)/config" />

    <!-- map_server -->
    <arg name="is_map" default="true" />
    <group if="$(arg is_map)" >
        <include file="$(find ros_utils)/launch/map_server.launch" />
    </group>

    <!-- rosbag -->
    <include file="$(find ros_utils)/launch/bag_player.launch" >
        <arg name="roomba"               value="$(arg roomba)" />
        <arg name="wait_time"            value="$(arg wait_time)" />
        <arg name="file_path"            value="$(env HOME)/bagfiles/experimental_file_$(arg date)/" />
        <arg name="convertor_theta_s"    value="true" />
        <arg name="publish_objects_data" value="false" />
    </include>

    <!-- tf_utils (only static tf) -->
    <include file="$(find ros_utils)/launch/rosbag_tf.launch" >
        <arg name="roomba"       value="$(arg roomba)"/>
        <arg name="publish_pose" value="false" />
    </include>

    <!-- object_localizer_node -->
    <arg name="publish_objects_data" default="false" />
    <node pkg="object_localizer" type="object_localizer_node" name="$(arg roomba)_object_localizer" output="screen" >
        <!-- remap -->
        <remap from="map_in"    to="map" />
        <remap from="odom_in"   to="$(arg roomba)/roomba/odometry" />
        <remap from="ops_in"    to="$(arg roomba)/object_positions" />
        <remap from="pose_out"  to="$(arg roomba)/pose" />
        <remap from="poses_out" to="$(arg roomba)/poses" />

        <!-- params -->
        <param name="BASE_LINK_FRAME_ID"   value="$(arg roomba)/base_link" />
        <param name="ROBOT_NAME"           value="$(arg roomba)" />
        <param name="IS_TF"                value="true" />
        <param name="PUBLISH_OBJECTS_DATA" value="$(arg publish_objects_data)" />
        <rosparam command="load" file="$(arg config_path)/params/object_localizer.yaml" />

        <!-- object_map -->
        <param name="INIT_OBJECTS_FILE_NAME" value="$(arg config_path)/init_objects/$(arg date).csv" />
        <rosparam command="load" file="$(arg config_path)/object_params/object_params.yaml" />
        
        <!-- dynamic_objects -->
        <rosparam command="load" file="$(arg config_path)/dynamic_objects/dynamic_objects.yaml" />
    </node>

    <!-- recorder -->
    <arg name="is_record" default="false" />
    <group if="$(arg is_record)" >
        <include file="$(find ros_utils)/launch/recorder.launch" >
            <arg name="roomba"                     value="$(arg roomba)"/>
            <arg name="is_record_pose"             value="true" />
	        <arg name="is_record_object_detection" value="true" />
	        <arg name="record_file_path"           value="$(find object_localizer)/record/$(arg roomba)_" />
        </include>
    </group>

    <!-- rviz -->
    <arg name="is_rviz" default="true" />
    <group if="$(arg is_rviz)" >
        <node pkg="rviz" type="rviz" name="$(arg roomba)_rviz" args="-d $(arg config_path)/rviz/object_localizer/$(arg roomba).rviz" />
    </group>

</launch>
